src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -423,16 +423,17 @@ private async ValueTask ProcessDataWrites()
                         {
                             // Output is ending and there are trailers to write
                             // Write any remaining content then write trailers
+
                             if (readResult.Buffer.Length > 0)
                             {
-                                // Only flush if required (i.e. content length exceeds flow control availability)
-                                // Writing remaining content without flushing allows content and trailers to be sent in the same packet
-                                await _frameWriter.WriteDataAsync(StreamId, _flowControl, readResult.Buffer, endStream: false, firstWrite, forceFlush: false);
+                                flushResult = await _frameWriter.WriteDataAndTrailersAsync(_stream, _flowControl, readResult.Buffer, firstWrite);
+                            }
+                            else
+                            {
+                                _stream.ResponseTrailers.SetReadOnly();
+                                _stream.DecrementActiveClientStreamCount();
```

Comment:
It would be simpler just to put the above two lines before this if/else and not try to put this logic inside _frameWriter.WriteDataAndTrailersAsync(). We normally call DecrementActiveClientStreamCount() before the last call to _frameWriter.WriteDataAsync(), so no real behavioral change there. That way, _frameWriter.WriteDataAndTrailersAsync() could take just the stream id and not the entire Http2Stream.